app/src/behaviors/behavior_tap_hold.c

```diff
@@ -0,0 +1,432 @@
+      k_msleep(10);
+    }
+    LOG_DBG("Releasing key position event for position %d %s", captured_position->position, (captured_position->state ? "pressed" : "released"));
+    ZMK_EVENT_RELEASE_AGAIN(captured_position);
+  }
+}
+
+
+/************************************************************  ACTIVE TAP HOLD HELPER FUNCTIONS */
+
+struct active_tap_hold* find_tap_hold(u32_t position) 
+{
+  for (int i = 0; i < ZMK_BHV_TAP_HOLD_MAX_HELD; i++) {
+    if (active_tap_holds[i].position == position) {
+      return &active_tap_holds[i];
+    }
+  }
+  return NULL;
+}
+
+struct active_tap_hold* store_tap_hold(u32_t position, const struct behavior_tap_hold_config* config) 
+{
+  for (int i = 0; i < ZMK_BHV_TAP_HOLD_MAX_HELD; i++) {
+    if (active_tap_holds[i].position != TH_POSITION_NOT_USED) {
+      continue;
+    }
+    active_tap_holds[i].position = position;
+    active_tap_holds[i].is_decided = false;
+    active_tap_holds[i].is_hold = false;
+    active_tap_holds[i].config = config;
+    return &active_tap_holds[i];
+  }
+  return NULL;
+}
+
+void clear_tap_hold(struct active_tap_hold * tap_hold) 
+{
+  tap_hold->position = TH_POSITION_NOT_USED;
+  tap_hold->is_decided = false;
+  tap_hold->is_hold = false;
+  tap_hold->work_is_cancelled= false;
+}
+
+enum decision_moment{
+  TH_KEY_UP = 0,
+  TH_OTHER_KEY_DOWN = 1,
+  TH_OTHER_KEY_UP = 2,
+  TH_TIMER_EVENT = 3,
+};
+
+static void decide_balanced(struct active_tap_hold * tap_hold, enum decision_moment event) {
+    switch(event) {
+    case TH_KEY_UP:
+      tap_hold->is_hold = 0;
+      tap_hold->is_decided = true;
+      break;
+    case TH_OTHER_KEY_UP:
+      tap_hold->is_hold = 1;
+      tap_hold->is_decided = true;
+      break;
+    case TH_TIMER_EVENT:
+      tap_hold->is_hold = 1;
+      tap_hold->is_decided = true;
+      break;
+    default: return;
+  }
+}
+
+static void decide_tap_preferred(struct active_tap_hold * tap_hold, enum decision_moment event) {
+    switch(event) {
+    case TH_KEY_UP:
+      tap_hold->is_hold = 0;
+      tap_hold->is_decided = true;
+      break;
+    case TH_TIMER_EVENT:
+      tap_hold->is_hold = 1;
+      tap_hold->is_decided = true;
+      break;
+    default: return;
+  }
+}
+
+static void decide_hold_preferred(struct active_tap_hold * tap_hold, enum decision_moment event) {
+    switch(event) {
+    case TH_KEY_UP: 
+      tap_hold->is_hold = 0;
+      tap_hold->is_decided = true;
+      break;
+    case TH_OTHER_KEY_DOWN:
+      tap_hold->is_hold = 1;
+      tap_hold->is_decided = true;
+      break;
+    case TH_TIMER_EVENT:
+      tap_hold->is_hold = 1;
+      tap_hold->is_decided = true;
+      break;
+    default: return;
+  }
+}
+
+
+void decide_tap_hold(struct active_tap_hold * tap_hold, enum decision_moment event)
+{
+  if (tap_hold->is_decided) {
+    return;
+  }
+
+  if(tap_hold != undecided_tap_hold) {
+    LOG_DBG("ERROR found undecided tap hold that is not the active tap hold");
+    return;
+  }
```

Comment:
this shouldn't happen ever, can we do something with asserts that are only built for debugging or something like that?